src/test/spec/retryable_reads.rs

```diff
@@ -0,0 +1,183 @@
+use bson::{doc, Bson, Document};
+
+use crate::test::{
+    assert_matches,
+    run_spec_test,
+    util::{parse_version, AnyTestOperation, EventClient, TestClient, TestEvent, TestFile},
+    CLIENT_OPTIONS,
+    LOCK,
+};
+
+const SKIPPED_OPERATIONS: &[&str] = &[
+    "count",
+    "download",
+    "download_by_name",
+    "listCollectionObjects",
+    "listDatabaseObjects",
+    "listIndexNames",
+    "listIndexes",
+    "mapReduce",
+    "watch",
+];
+
+#[cfg_attr(feature = "tokio-runtime", tokio::test)]
+#[cfg_attr(feature = "async-std-runtime", async_std::test)]
+async fn run() {
+    async fn run_test(test_file: TestFile) {
+        let has_skipped_op = test_file.tests.iter().any(|test_case| {
+            test_case
+                .operations
+                .iter()
+                .any(|op| SKIPPED_OPERATIONS.contains(&op.name.as_str()))
+        });
+        if has_skipped_op {
+            return;
+        }
+
+        for test_case in test_file.tests {
+            if let Some(skip_reason) = test_case.skip_reason {
+                println!("Skipping {}: {}", test_case.description, skip_reason);
+                continue;
+            }
+
+            let mut options = CLIENT_OPTIONS.clone();
+            if let Some(client_options) = test_case.client_options {
+                options.retry_reads = Some(client_options.get_bool("retryReads").unwrap());
+            }
+            if TestClient::new().await.is_sharded() && test_case.use_multiple_mongoses != Some(true)
+            {
+                options.hosts = options.hosts.iter().cloned().take(1).collect();
+            }
+            let client = EventClient::with_options(options).await;
+
+            if let Some(ref run_on) = test_file.run_on {
+                let can_run_on = run_on.iter().any(|run_on| {
```

Comment:
this logic can be nicely abstracted to a method on `RunOn`, which will be useful for subsequent runners.